paasta_tools/kubernetes_test_lib/utils.py

```diff
@@ -0,0 +1,66 @@
+from collections import defaultdict
+
+from kubernetes.client import V1Deployment
+from kubernetes.client import V1DeploymentList
+from kubernetes.client import V1StatefulSet
+from kubernetes.client import V1StatefulSetList
+from kubernetes.client.rest import ApiException
+
+from paasta_tools.kubernetes_tools import KubeClient
+
+
+class MockApi():
+    """
+    This is a very very simple object that mocks Kubernetes client.
+    Be sure to add/update it before using it.
+    """
+    deployments = defaultdict(dict)
+    stateful_sets = defaultdict(dict)
+    pod_disruption_budgets = defaultdict(dict)
+
+    def list_namespaced_deployment(self, namespace, *args):
+        res = V1DeploymentList(items=[v for k, v in self.deployments[namespace].items()])
+        return res
+
+    def list_namespaced_stateful_set(self, namespace, *args):
+        res = V1StatefulSetList(items=[v for k, v in self.stateful_sets[namespace].items()])
+        return res
+
+    def create_namespaced_deployment(self, namespace, body, *args):
+        if not isinstance(body, V1Deployment):
+            raise Exception('body should be V1Deployment')
+        self.deployments[namespace][body.metadata.name] = body
+
+    def create_namespaced_stateful_set(self, namespace, body, *args):
+        if not isinstance(body, V1StatefulSet):
+            raise Exception('body should be V1Deployment')
+        self.stateful_sets[namespace][body.metadata.name] = body
+
+    def delete_namespaced_deployment(self, name, namespace, *args):
+        if namespace not in self.deployments or name not in self.deployments[namespace]:
+            raise ApiException(status=404, reason='name or namespace does not exist')
+        self.deployments[namespace].pop(name, None)
+
+    def delete_namespaced_stateful_set(self, name, namespace, *args):
+        if namespace not in self.stateful_sets or name not in self.stateful_sets[namespace]:
+            raise ApiException(status=404, reason='name or namespace does not exist')
+        self.stateful_sets[namespace].pop(name, None)
+
+    def delete_namespaced_pod_disruption_budget(self, name, namespace, *args, **kwargs):
+        if namespace not in self.pod_disruption_budgets or name not in self.pod_disruption_budgets[namespace]:
+            raise ApiException(status=404, reason='name or namespace does not exist')
+        self.pod_disruption_budgets[namespace].pop(name, None)
+
+
+class MockKubeClient(KubeClient):
+    """
+    This is a very very simple object that mocks KubeClient
+    Be sure to add/update it before using it.
+    """
+
+    def __init__(self):
```

Comment:
Are you sure we need this? Is this better than having a generic mock that we assert calls on? 

This is already different than the real KubeClient because you're assuming the same API methods for everything (e.g. `self.deployments`, `self.core`) and I don't see much benefit from having to keep the mock API objects up to date.

Reply:
i was thinking about using this file as a library, and add more functions to it in the future. Don't I have to patch all function calls if I use a generic mock? And a generic mock won't save the objects. I mostly want to avoid writing a bunch of mock.patch. 
We could easily add more functionality to this file if we need it in the future, since it's extensible.

Reply:
You don't have to patch all the function calls, any attribute of a mock is also a mock. So `mock_client.deployments.list_namespaced_deployment` would be a mock that you can make assertions on. It won't save the objects, but when you have that logic in the mock, it's less clear what's happening in the test, and you run the risk of the mock logic being different than the real API.